### PR TITLE
🐛 persist dynamic-sidecar state to docker service labels

### PR DESCRIPTION
**Pull request by GitHK**
## What do these changes do?

Previously only partial information was stored on the labels attached to the dynamic-sidecar. Now all the information tracked by `SchedulerData` is stored inside the `io.osparc.scheduler-data` label. This label is updated every the `SchedulerData` changes.

Extras:
- dynamic-sidecar mounted volumes are now dependent on the deployment `docker_swarm_name` label
- fixed regex for DockerID, it was previously tiring to validate a SHA256 

## Checklist

- [x] Unit tests for the changes exist
- [x] Runs in the swarm

### CODE REVIEW

File: services/director-v2/src/simcore_service_director_v2/models/schemas/dynamic_services/scheduler.py [hunk 1]
```diff
@@ -28,9 +27,9 @@
 
 MAX_ALLOWED_SERVICE_NAME_LENGTH: int = 63
 
-SHA256 = constr(max_length=64, regex=r"\b[A-Fa-f0-9]{64}\b")
-ServiceId = SHA256
-NetworkId = SHA256
+DockerId = constr(max_length=25, regex=r"[A-Za-z0-9]{25}")
```
Comment (pcrespov) [inline services/director-v2/src/simcore_service_director_v2/models/schemas/dynamic_services/scheduler.py:30]: nice

File: services/director-v2/src/simcore_service_director_v2/models/schemas/dynamic_services/scheduler.py [hunk 2]
```diff
@@ -432,42 +370,16 @@ def from_http_request(
...
-                request_scheme=request_scheme,
-                proxy_service_name=proxy_service_name,
-                dynamic_sidecar=dict(
-                    hostname=service_labels_stored_data.service_name,
-                    port=port,
-                ),
-            )
-        )
+        labels = service_inspect["Spec"]["Labels"]
+        return cls.parse_obj(json.loads(labels[DYNAMIC_SIDECAR_SCHEDULER_DATA_LABEL]))
```
Comment (pcrespov) [inline services/director-v2/src/simcore_service_director_v2/models/schemas/dynamic_services/scheduler.py:377]: TIP: you can use ``cls.parse_raw(labels[DYNAMIC_SIDECAR_SCHEDULER_DATA_LABEL])`` here

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api.py [hunk 3]
```diff
@@ -506,3 +511,39 @@ async def try_to_remove_network(network_name: str) -> None:
...
+
+        try:
+            # fetch information from service name
+            service_inspect = await client.services.inspect(scheduler_data.service_name)
+            service_version = service_inspect["Version"]["Index"]
+            service_id = service_inspect["ID"]
+            spec = service_inspect["Spec"]
+
+            # allows to use json encodes not available on dict
+            dict_scheduler_data = json.loads(scheduler_data.json())
```
Comment (pcrespov) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api.py:529]: We have discussed about this many times now ... please check and use what is inside ``models_library.utils.fastapi_encoders``

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api.py [hunk 4]
```diff
@@ -506,3 +511,39 @@ async def try_to_remove_network(network_name: str) -> None:
...
+            # compose_spec needs to be json encoded
+            dict_scheduler_data["compose_spec"] = json.dumps(
```
Comment (pcrespov) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api.py:531]: we should elaborate on this at some point... since we have the json-schema of the compose-specs and I think it would make sense to use a pydantic model of the compose-specs as well.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api.py [hunk 5]
```diff
@@ -506,3 +511,39 @@ async def try_to_remove_network(network_name: str) -> None:
...
+                dict_scheduler_data["compose_spec"]
+            )
+            label_data = json.dumps(dict_scheduler_data)
+            spec["Labels"][DYNAMIC_SIDECAR_SCHEDULER_DATA_LABEL] = label_data
```
Comment (pcrespov) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api.py:535]: I do not understand this. Are you modifying here the image labels?

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api.py [hunk 6]
```diff
@@ -506,3 +511,39 @@ async def try_to_remove_network(network_name: str) -> None:
...
+
+            await client._query_json(  # pylint: disable=protected-access
```
Comment (pcrespov) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api.py:537]: using a protected member of a third part might have compatibility issues in a near future. There is really no other way to update a service?

MINOR: ``f"services/{service_id}/update"`` is shorter

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api.py [hunk 7]
```diff
@@ -506,3 +511,39 @@ async def try_to_remove_network(network_name: str) -> None:
...
+                "services/{service_id}/update".format(service_id=service_id),
+                method="POST",
+                data=json.dumps(clean_map(spec)),
+                params={"version": service_version},
+            )
+        except aiodocker.exceptions.DockerError as e:
+            if e.message != f"service {scheduler_data.service_name} not found":
```
Comment (pcrespov) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api.py:544]: Using a message is also quite unreliable way ot determine a state. 
I see that https://aiodocker.readthedocs.io/en/latest/_modules/aiodocker/exceptions.html#DockerError provides a status flag. If those are http status flags, then e.status==404 might be more reliable.

File: services/director-v2/tests/unit/with_swarm/test_dynamic_sidecar_scheduler_docker_api.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
+# pylint: disable=redefined-outer-name
+# pylint: disable=unused-argument
+
```
Comment (pcrespov) [inline services/director-v2/tests/unit/with_swarm/test_dynamic_sidecar_scheduler_docker_api.py:3]: The name of this test does no t match the path of the python module it tests

``services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api.py`` -->
``test_modules_dynamic_sidecar_docker_api.py``

File: services/director-v2/tests/unit/with_swarm/test_dynamic_sidecar_scheduler_docker_api.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        DockerContainerInspect.from_container(
+            {"State": {"Status": "dead"}, "Name": "mock_name", "Id": "mock_id"}
+        )
+    ]
+    scheduler_data.service_name = service_name
+    return scheduler_data
+
+
+@pytest.fixture
+async def docker() -> AsyncIterable[aiodocker.Docker]:
```
Comment (pcrespov) [inline services/director-v2/tests/unit/with_swarm/test_dynamic_sidecar_scheduler_docker_api.py:60]: don't we have this fixtures somehere else? Should it not be in pytest-simcore?

File: services/director-v2/tests/unit/with_swarm/test_dynamic_sidecar_scheduler_docker_api.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    mock_service: str,
+    mock_scheduler_data: SchedulerData,
+    docker_swarm: None,
+) -> None:
+    await update_scheduler_data_label(mock_scheduler_data)
+
+    # fetch stored data in labels
+    service_inspect = await docker.services.inspect(mock_service)
+    labels = service_inspect["Spec"]["Labels"]
+    scheduler_data = SchedulerData.parse_obj(
```
Comment (pcrespov) [inline services/director-v2/tests/unit/with_swarm/test_dynamic_sidecar_scheduler_docker_api.py:90]: TIP: can use ``parse_raw``

File: services/director-v2/tests/unit/test_schemas_dynamic_services_scheduler.py [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+        target_prop=lambda x: x.dynamic_sidecar.service_removal_state,
+        attr_name="was_removed",
+        value=True,
+    ),
+]
+
+
+@pytest.mark.parametrize("change", CHANGES, ids=[x.attr_name for x in CHANGES])
+async def test_nested_compare(fake_scheduler_data: str, change: Change) -> None:
+    scheduler_data = SchedulerData.parse_obj(json.loads(fake_scheduler_data))
```
Comment (pcrespov) [inline services/director-v2/tests/unit/test_schemas_dynamic_services_scheduler.py:72]: TIP: parse_raw

File: services/director-v2/tests/unit/test_schemas_dynamic_services_scheduler.py [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+    original_scheduler_data = deepcopy(scheduler_data)
```
Comment (pcrespov) [inline services/director-v2/tests/unit/test_schemas_dynamic_services_scheduler.py:73]: TIP: pydantic also implements a copy member SEE https://pydantic-docs.helpmanual.io/usage/exporting_models/#modelcopy

File: services/director-v2/tests/unit/test_schemas_dynamic_services_scheduler.py [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+
+    # change change to attribute
+    change.target_prop(scheduler_data).__setattr__(change.attr_name, change.value)
```
Comment (pcrespov) [inline services/director-v2/tests/unit/test_schemas_dynamic_services_scheduler.py:76]: if i understand it correctly, what you intend to do here is something like
```python
scheduler_data.some_field["some_attr"] = some_value
```
where ``some_field, some_attr, some_value`` are given by the parametrization, right?

For a test, i find this very convoluted to read and I would have used a more verbatim approach

```python
scheduler_data.dynamic_sidecar.status["info"] = "some info"
assert original_scheduler_data != scheduler_data

scheduler_data.dynamic_sidecar.service_removal_state["was_removed"] = True
assert original_scheduler_data != scheduler_data

# etc
```

TIP: check python [``operator`` module and ``.attrgetter, itemgetter, attrsetter, itemgetter``](https://docs.python.org/3/library/operator.html#operator.attrgetter) ... might make it more readable

**Review by pcrespov — COMMENTED** [review]
Q: does this has any effect on the image labels?

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api.py (hunk 6 above)
Comment (GitHK) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api.py:537]: Unfortunately there is no other way, but I have tests in place to guarantee this does not break.

File: services/director-v2/tests/unit/with_swarm/test_dynamic_sidecar_scheduler_docker_api.py (hunk 9 above)
Comment (GitHK) [reply services/director-v2/tests/unit/with_swarm/test_dynamic_sidecar_scheduler_docker_api.py:60]: We have something similar with `python docker` not `aiodocker`

File: services/director-v2/tests/unit/test_schemas_dynamic_services_scheduler.py (hunk 13 above)
Comment (GitHK) [reply services/director-v2/tests/unit/test_schemas_dynamic_services_scheduler.py:76]: I just managed to make something more readable using a context manger. Let's see if you like it better like that. Also looks simpler.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api.py (hunk 5 above)
Comment (GitHK) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api.py:535]: Yes I am overwriting a label, the one which keeps track of the state of the service.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api.py (hunk 4 above)
Comment (GitHK) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api.py:531]: I agree with you. You already mentioned this. I'm planning to have a PR for this in my backlog.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api.py (hunk 7 above)
Comment (GitHK) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api.py:544]: I think that adding `e.status!=404` to the current check would be even better.

I don't see how this is less reliable. I am searching for a very specific message, corresponding to a very specific case.

File: services/director-v2/tests/unit/test_schemas_dynamic_services_scheduler.py (hunk 12 above)
Comment (GitHK) [reply services/director-v2/tests/unit/test_schemas_dynamic_services_scheduler.py:73]: Good to know about. But will go with the tried and tested standardlib function, in this situation I don't need to exclude or include other fields. `deepcopy` is perfect.

**Review by pcrespov — APPROVED** [review]

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api.py (hunk 3 above)
Comment (GitHK) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api.py:529]: Thanks for pointing this out. It would be helpful in the future.
For now I've properly fixed the code here. it should be much better.

File: services/director-v2/src/simcore_service_director_v2/core/settings.py [hunk 14]
```diff
@@ -221,6 +221,13 @@ class DynamicSidecarSettings(BaseCustomSettings):
             "timeout for attaching/detaching project networks to/from a container"
         ),
     )
+    DYNAMIC_SIDECAR_VOLUMES_REMOVAL_TIMEOUT_S: PositiveFloat = Field(
+        1.0 * MINS,
+        description=(
+            "time to wait before giving up on removing dynamic-sidecar's volumes "
+            "used for syncing data for: inputs, outputs and state paths"
```
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/core/settings.py:228]: MINOR: this line is unnecessary. these are internals.

File: services/director-v2/src/simcore_service_director_v2/models/schemas/constants.py [hunk 15]
```diff
@@ -3,6 +3,10 @@
 DYNAMIC_SIDECAR_SERVICE_PREFIX = "dy-sidecar"
 DYNAMIC_PROXY_SERVICE_PREFIX = "dy-proxy"
 
+# label storing scheduler_data to allow service
+# monitoring recovery after director-v2 reboots
+DYNAMIC_SIDECAR_SCHEDULER_DATA_LABEL = "io.osparc.scheduler-data"
```
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/models/schemas/constants.py:8]: I do have a question here. Why ```io.osparc.```?
all the other labels are using ```io.simcore``` why this change?

File: services/director-v2/src/simcore_service_director_v2/models/schemas/dynamic_services/scheduler.py [hunk 16]
```diff
@@ -432,38 +369,11 @@ def from_http_request(
...
-                service_port=service_labels_stored_data.service_port,
-                request_dns=request_dns,
-                request_scheme=request_scheme,
-                proxy_service_name=proxy_service_name,
-                dynamic_sidecar=dict(
-                    hostname=service_labels_stored_data.service_name,
-                    port=port,
-                ),
-            )
-        )
```
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/models/schemas/dynamic_services/scheduler.py:466]: :+1:  I like it when things are deleted...

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_service_specs/sidecar.py [hunk 17]
```diff
@@ -187,13 +191,8 @@ def get_dynamic_sidecar_spec(
...
             "uuid": f"{scheduler_data.node_uuid}",  # also needed for removal when project is closed
-            "swarm_stack_name": dynamic_sidecar_settings.SWARM_STACK_NAME,
-            "service_key": scheduler_data.key,
-            "service_tag": scheduler_data.version,
-            "paths_mapping": scheduler_data.paths_mapping.json(),
-            "compose_spec": json_dumps(scheduler_data.compose_spec),
-            "container_http_entry": scheduler_data.container_http_entry,
-            "restart_policy": scheduler_data.restart_policy,
+            "swarm_stack_name": dynamic_sidecar_settings.SWARM_STACK_NAME,  # required for listing services with uuid
+            DYNAMIC_SIDECAR_SCHEDULER_DATA_LABEL: scheduler_data.json(),
```
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_service_specs/sidecar.py:195]: nice!

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/task.py [hunk 18]
```diff
@@ -285,6 +290,7 @@ async def retrieve_service_inputs(
 
         return RetrieveDataOutEnveloped.from_transferred_bytes(transferred_bytes)
 
+    # NOTE: used by API indirectly
```
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/task.py:293]: why all these comments? also be aware that ```isort``` sometimes moves these at the top of the file... so I'm not sure it's such a great idea

File: services/director-v2/tests/mocks/fake_scheduler_data.json [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+    "user_id": 1,
+    "project_id": "c440556a-b4d9-11ec-9a17-02420a000016",
+    "node_uuid": "3e68d1f6-be3e-414e-a468-4a2bf415f756",
+    "service_name": "dy-sidecar_3e68d1f6-be3e-414e-a468-4a2bf415f756",
+    "dynamic_sidecar": {
+        "status": {
+            "current": "ok",
+            "info": ""
+        },
+        "hostname": "dy-sidecar_3e68d1f6-be3e-414e-a468-4a2bf415f756",
```
Comment (sanderegg) [inline services/director-v2/tests/mocks/fake_scheduler_data.json:24]: simple question: what happens if the node where the sidecar was goes down? is the service moving to another one? will it fail?

**Review by sanderegg — COMMENTED** [review]
Cool! please look into why the label is called ```io.osparc``` instead of ```io.simcore```. To me this sounds the same but not really. I would like to understand the rationale first.

File: services/director-v2/tests/mocks/fake_scheduler_data.json (hunk 19 above)
Comment (GitHK) [reply services/director-v2/tests/mocks/fake_scheduler_data.json:24]: Some notes:
- When a dynamic-sidecar is started is paired together with a proxy which is scheduled to the same node where the dynamic-sidecar is running.
- Since moving the proxy from Traefik to Caddy, it is no longer a requirement to be on the same node.
- If the service gets removed, all its spawned containers are are also removed. So cleanup is OK.

Now if the node on which the dynamic-sidecar is running goes down the director-v2 will no longer see the service can no longer contact it. Will mark is as unhealthy and try to remove it from tracking. Will have to check this.

File: services/director-v2/src/simcore_service_director_v2/models/schemas/constants.py (hunk 15 above)
Comment (GitHK) [reply services/director-v2/src/simcore_service_director_v2/models/schemas/constants.py:8]: confused label, will use `io.simcore`

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/task.py (hunk 18 above)
Comment (GitHK) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/task.py:293]: Let me just drop them. If they cause issues.
